cpp/daal/src/data_management/train_test_split.cpp

```diff
@@ -39,6 +41,117 @@ typedef daal::data_management::NumericTable::StorageLayout NTLayout;
 const size_t BLOCK_CONST      = 2048;
 const size_t THREADING_BORDER = 8388608;
 
+size_t genSwapIdx(size_t i, unsigned int * randomNumbers, size_t & rnIdx)
+{
+    uint32_t bitMask = i;
+    bitMask |= bitMask >> 1;
+    bitMask |= bitMask >> 2;
+    bitMask |= bitMask >> 4;
+    bitMask |= bitMask >> 8;
+    bitMask |= bitMask >> 16;
+
+    size_t j = randomNumbers[rnIdx] & bitMask;
+    while (j > i)
+    {
+        ++rnIdx;
+        j = randomNumbers[rnIdx] & bitMask;
+    }
+    ++rnIdx;
+    return j;
+}
+
+template <daal::CpuType cpu>
+services::Status generateRandomNumbers(const int * rngState, unsigned int * randomNumbers, const size_t nSkip, const size_t n)
+{
+    daal::internal::mkl::BaseRNG<cpu> baseRng(0, VSL_BRNG_MT19937);
+    daal::services::internal::TArray<unsigned int, cpu> baseRngStateArr(baseRng.getStateSize() / 4);
+    unsigned int * baseRngState = baseRngStateArr.get();
+    DAAL_CHECK_MALLOC(baseRngState);
+    baseRng.saveState(baseRngState);
+    for (size_t i = 4; i < 628; ++i) baseRngState[i] = rngState[i - 4];
+    baseRngState[628] = 624;
+    baseRng.loadState(baseRngState);
+    daal::internal::RNGs<unsigned int, cpu> rng;
+
+    if (nSkip != 0) baseRng.skipAhead(nSkip);
+
+    daal::services::internal::TArray<unsigned int, cpu> tmpArr(n + 16);
+    unsigned int * tmp = tmpArr.get();
+    DAAL_CHECK_MALLOC(tmp);
+    rng.uniformBits32(n + 16, tmp, baseRng.getState());
+
+    for (size_t i = 0; i < n; ++i) randomNumbers[nSkip + i] = tmp[i];
```

Comment:
I think it would be better to generate `n` elements right to `randomNumbers` buffer without copy, and than generate 16 elements to `tmp` (tmpArr(16)). 
**And it's still not clear**: why we generates `n+16` but uses only first `n` elements?

Reply:
It doesn't work so: if number of elements = n, then last 16 of them will be different than numpy's. Reason of that is unclear and requiring investigation with deep dive in RNG work, to save time it's better to leave code as is.